### PR TITLE
Add mean to std.algorithm

### PR DESCRIPTION
**Pull request by JackStouffer**
Split from https://github.com/D-Programming-Language/phobos/pull/3592

I will propose the same question I had in the other PR

>I also want to add various other statistical methods to Phobos, such as standard deviation, which fit in std.algorithm, and things such as rolling mean, which don't fit into std.algorithm. I was thinking that something like a rolling mean would best fit in std.numeric, but, would it be best if all of the statistical methods were in the same module, or is it ok to have them separate?

Also fixes issue 14034

### CODE REVIEW

File: std/algorithm/iteration.d [hunk 1]
```diff
@@ -4284,3 +4286,112 @@ unittest
...
+    if (r.empty)
+    {
+        return real.nan;
+    }
+
+    return mean(r, real(0.0));
+}
+
+/// ditto
+auto mean(R, T)(R r, T seed)
```
Comment (burner) [inline std/algorithm/iteration.d]: as this function is not used directly in your tests I suppose it should be private and named meanImpl

Comment (JackStouffer) [inline std/algorithm/iteration.d]: @burner See https://github.com/D-Programming-Language/phobos/pull/3679/files#diff-0d6463fc6f41c5fb774300832e3135f5R4395

Comment (burner) [inline std/algorithm/iteration.d]: my bad

**Comment by 9il** [comment]
Hi @JackStouffer!
I would like to see more PRs like that :-)
Few notes:

1. This and any other statistical PRs should depend on #2991  
2. We need to discuss module names first. A submodule for `std.numeric` is more suitable, thought. We should save `std.algorithm.iteration` as general purpose module without serious math payload.
2. Please ping @John-Colvin and @9il (me) when you create math PRs.

**Comment by JackStouffer** [comment]
> This and any other statistical PRs should depend on #2991

Ok, I'm not quite sure how to make a branch off of a PR though.

> We need to discuss module names first. A submodule for std.numeric is more suitable, though

`std.numeric.stats` perhaps? If that's the case, then your `geometricMean` function should be moved there as well.

> We should save `std.algorithm.iteration` as general purpose module without serious math payload.

I agree. Following this, I believe the `sum` functions in `std.algorithm.iteration` should be deprecated aliases and move the actual functions in `std.numeric` then. That way all of the math code is in one place.

**Comment by 9il** [comment]
> Ok, I'm not quite sure how to make a branch off of a PR though.

You can fork my phobos repo and switch to the corresponding branch. Please read last comments starting with https://github.com/D-Programming-Language/phobos/pull/2991#issuecomment-141816906 . Summation API can be changed during October. You are welcome for API review.

> std.numeric.stats perhaps? If that's the case, then your geometricMean function should be moved there as well.

`std.numeric.stats` sounds good. For first release it may be small DUB package and `std.experimental.numeric.stats`.

**Comment by DmitryOlshansky** [comment]
Glad to see this is converging on nice package names and goes in the right direction ;)

**Comment by JackStouffer** [comment]
https://github.com/9il/phobos/pull/3

**Comment by JackStouffer** [comment]
Now that https://github.com/D-Programming-Language/phobos/pull/2991 has been closed, I am reopening this PR. I have moved mean to std.numeric as I believe that it is a better fit for this function than std.algorithm.

**Pull request closed without merging**